enterprise/mqtt/src/main/java/io/crate/mqtt/operations/MqttIngestService.java

```diff
@@ -0,0 +1,267 @@
+        new QualifiedName("topic"), 2,
+        new QualifiedName("ts"), 3,
+        new QualifiedName("payload"), 4);
+    private static final List<DataType> FIELD_TYPES = Arrays.asList(DataTypes.STRING, DataTypes.INTEGER, DataTypes.STRING, DataTypes.OBJECT);
+    private static final Predicate<Row> ALWAYS_TRUE = (r) -> true;
+
+    private final ExpressionAnalyzer expressionAnalyzer;
+    private final InputFactory inputFactory;
+    private final SQLOperations sqlOperations;
+    private final IngestionService ingestionService;
+    private final User crateUser;
+    private final AtomicReference<Set<Tuple<Predicate<Row>, IngestRule>>> predicateAndIngestRulesReference =
+        new AtomicReference<>(new HashSet<>());
+    private final ExpressionAnalysisContext expressionAnalysisContext;
+    private boolean isInitialized;
+
+    public MqttIngestService(Functions functions,
+                             SQLOperations sqlOperations,
+                             UserLookup userLookup,
+                             IngestionService ingestionService) {
+        this.sqlOperations = sqlOperations;
+        this.inputFactory = new InputFactory(functions);
+        this.expressionAnalysisContext = new ExpressionAnalysisContext();
+        FieldProvider<Symbol> mqttSourceFieldsProvider = new FieldProvider<Symbol>() {
+
+            @Override
+            public Symbol resolveField(QualifiedName qualifiedName, Operation operation) {
+                return resolveField(qualifiedName, null, operation);
+            }
+
+            @Override
+            public Symbol resolveField(QualifiedName qualifiedName, @Nullable List<String> path, Operation operation) {
+                return new InputColumn(MQTT_FIELDS_ORDER.get(qualifiedName));
+            }
+        };
+        this.expressionAnalyzer = new ExpressionAnalyzer(functions, null, null, mqttSourceFieldsProvider, null);
+        this.ingestionService = ingestionService;
+        this.crateUser = userLookup.findUser("crate");
+    }
+
+    /**
+     * Registers the service as an #{@link IngestRuleListener} with the #{@link IngestionService}.
+     * This must only be called once.
+     */
+    public void initialize() {
+        if (isInitialized) {
+            throw new IllegalStateException("Service already initialized");
+        }
+        ingestionService.registerIngestRuleListener(SOURCE_IDENT, this);
+        isInitialized = true;
+    }
+
+    private static String mqttPayloadToString(ByteBuf content) {
+        int size = content.readableBytes();
+        byte[] rawBytes = new byte[size];
+        content.getBytes(content.readerIndex(), rawBytes);
+        return new String(rawBytes, Charsets.UTF_8);
+    }
+
+    @Nullable
+    private static Map<String, Object> parsePayloadToMap(ByteBuf payload) {
+        String json = mqttPayloadToString(payload);
+        try {
+            return JsonXContent.jsonXContent.createParser(NamedXContentRegistry.EMPTY, json).mapOrdered();
```

Comment:
I think the ordering doesn't matter here? Changing this from `mapOrdered()` to `map()` would be cheaper (uses a `HashMap` instead of LinkedHashMap`)